packages/tre/src/plugins/core/errors/index.ts

```diff
@@ -196,36 +196,73 @@ function getSourceMappedStack(workerSrcOpts: SourceOptions[], error: Error) {
 // matching the schema below with the `MF-Experimental-Error-Stack` header set
 // to a truthy value, in order to display the pretty-error page.
 export const HEADER_ERROR_STACK = "MF-Experimental-Error-Stack";
-const ErrorSchema = z.object({
-  message: z.ostring(),
-  name: z.ostring(),
-  stack: z.ostring(),
-});
+
+export interface JsonError {
+  message?: string;
+  name?: string;
+  stack?: string;
+  cause?: JsonError;
+}
+export const JsonErrorSchema: z.ZodType<JsonError> = z.lazy(() =>
+  z.object({
+    message: z.ostring(),
+    name: z.ostring(),
+    stack: z.ostring(),
+    cause: JsonErrorSchema.optional(),
+  })
+);
+
+export function reviveError(
+  workerSrcOpts: SourceOptions[],
+  jsonError: JsonError
+): Error {
+  let cause: Error | undefined;
+  if (jsonError.cause !== undefined) {
+    cause = reviveError(workerSrcOpts, jsonError.cause);
+  }
+
+  // If this is one of hte built-in error types, construct an instance of that
+  let ctor = Error;
+  if (jsonError.name !== undefined && jsonError.name in globalThis) {
+    const maybeCtor = (globalThis as Record<string, unknown>)[jsonError.name];
+    if (Object.getPrototypeOf(maybeCtor) === Error) {
+      ctor = maybeCtor as typeof Error;
+    }
+  }
+
+  // Construct the error, copying over the correct name and stack trace
+  const error = new ctor(jsonError.message, { cause });
+  if (jsonError.name !== undefined) error.name = jsonError.name;
```

Comment:
Why is this line needed? If `jsonError.name !== undefined` then presumably the error was constructed with the right class to get the correct name? (Also see above re allow listing)